### PR TITLE
[FEAT]: add partitioning_* functions to sql

### PR DESCRIPTION
**Pull request by universalmind303**

### CODE REVIEW

**Comment by kevinzwang** [comment]
@universalmind303 could you wait until https://github.com/Eventual-Inc/Daft/pull/2842 to merge this? I made a few changes to our partitioning functions

File: src/daft-sql/src/modules/partitioning.rs [hunk 1]
```diff
@@ -1,11 +1,96 @@
...
+                partitioning_helper(args, planner, "months", partitioning::months)
+            }
+            PartitioningExpr::Days => {
+                partitioning_helper(args, planner, "days", partitioning::days)
+            }
+            PartitioningExpr::Hours => {
+                partitioning_helper(args, planner, "hours", partitioning::hours)
+            }
+            PartitioningExpr::IcebergBucket(_) => {
+                let input = planner.plan_function_arg(&args[0])?;
```
Comment (raunakab) [inline src/daft-sql/src/modules/partitioning.rs:58]: Could `args` potentially be an empty slice here? And therefore, `&args[0]` would fail.

Comment (raunakab) [reply src/daft-sql/src/modules/partitioning.rs:58]: (Unless we're checking before hand that `args` is not empty for the `IcebergBucket` case).

**Review by raunakab — COMMENTED** [review]
Just a question about length checking for specific cases.

File: src/daft-sql/src/modules/partitioning.rs (hunk 1 above)
Comment (universalmind303) [reply src/daft-sql/src/modules/partitioning.rs:58]: good catch, we should definitely check the num args first!

**Pull request merged**